examples/seismic/acoustic/gradient_example.py

```diff
@@ -1,109 +1,147 @@
 import numpy as np
 from numpy import linalg
+from cached_property import cached_property
 
 from devito import TimeFunction, Function
 from examples.seismic import Receiver, RickerSource, demo_model
 from examples.seismic.acoustic import ForwardOperator, GradientOperator, smooth10
 
+class GradientExample(object):
+    def __init__(self, shape=(50, 50, 50), spacing=(15.0, 15.0, 15.0), tn=500., time_order=2,
+            space_order=4, nbpml=10):
+        self.time_order = time_order
+        self.space_order = space_order
+        self._setup_model_and_acquisition(shape, spacing, nbpml, tn)
+        self._true_data()
+
+    @cached_property
+    def dt(self):
+        return self.model.critical_dt * (1.73 if self.time_order == 4 else 1.0)
+
+    def _setup_model_and_acquisition(self, shape, spacing, nbpml, tn):
```

Comment:
nitpicking: this could/should be a free function, returning whatever is needed for initialisation

Reply:
If you see the previous version (the one in master currently) of this file, it does what (I think) you're looking for. Since there are so many objects that are created during initialisation, having a free function here means having ~ 10 initialised objects being returned and then being passed around for the other method calls. Having them as class properties to reduce the parameter clutter was a big part of this cleanup.